middleware/proxy/proxy.go

```diff
@@ -60,6 +69,60 @@ func (uh *UpstreamHost) Down() bool {
 	return uh.CheckDown(uh)
 }
 
+func (uh *UpstreamHost) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg, p Proxy) (*dns.Msg, error) {
+
+	var reply *dns.Msg
+	var backendErr error
+
+	atomic.AddInt64(&uh.Conns, 1)
+
+	switch uh.encoding {
+		case encodingNone:
+			reply, backendErr = p.Client.ServeDNS(w, r, uh)
+		case encodingGRPC:
+			reply, backendErr = uh.serveGRPC(ctx, w, r)
+		case encodingGRPCTLS:
+			reply, backendErr = uh.serveGRPC(ctx, w, r)
+		default:
+			reply, backendErr =  nil, errInvalidEncoding
+	}
+	atomic.AddInt64(&uh.Conns, -1)
+
+	return reply, backendErr
+}
+
+func (uh *UpstreamHost) serveGRPC(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (*dns.Msg, error) {
+	if uh.grpc == nil {
+		var conn *grpc.ClientConn
+		var err error
+		if uh.tls != nil {
+			conn, err = grpc.Dial(uh.Name, grpc.WithTransportCredentials(creds.NewTLS(uh.tls)))
```

Comment:
So this redials for every new connection? I.e. this is a new setup of a TCP conn? (And then turning is down again)?